ext/drbbdup/drbbdup.c

```diff
@@ -423,20 +441,14 @@ static dr_emit_flags_t
 drbbdup_duplicate_phase(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
                         bool translating)
 {
-    if (translating)
-        return DR_EMIT_DEFAULT;
-
-    app_pc pc = instr_get_app_pc(instrlist_first_app(bb));
-    ASSERT(pc != NULL, "pc cannot be NULL");
-
     dr_rwlock_write_lock(rw_lock);
 
     drbbdup_manager_t *manager =
-        (drbbdup_manager_t *)hashtable_lookup(&manager_table, pc);
+        (drbbdup_manager_t *)hashtable_lookup(&manager_table, tag);
 
-    if (!for_trace && manager != NULL && !manager->is_gen) {
+    if (!for_trace && !translating && manager != NULL && !manager->is_gen) {
         /* Remove existing invalid book-keeping data. */
-        hashtable_remove(&manager_table, pc);
+        hashtable_remove(&manager_table, tag);
```

Comment:
Yes it is trying to handle app code mods.

Reply:
> What about thread-private caches: this will throw away on any block duplicated among threads?

We assume global cache for now. drbbdup does not support thread-private caches well atm.

Reply:
@derekbruening By any chance, do you have a better idea for detecting app code mods?

Reply:
> @derekbruening By any chance, do you have a better idea for detecting app code mods?

Having something like #409 would be good.  We don't have that today though, and with races limiting dr_register_delete_event(), just looking for duplicate tags is a reasonable approach.  IIRC Dr. Memory is doing something like this too.  Its comment says it handle thread-private; haven't re-examined it enough to explain that well.  https://github.com/DynamoRIO/drmemory/blob/master/drmemory/spill.c#L1755